### ISSUE TITLE
Configurable Product Links Validation Bug when int value is 0

### ISSUE THREAD

**Issue by davidverholen** [magento/magento2#13210]
### Preconditions
1. Magento 2.4-develop
2. Attribute with Source Model having 0 as one value
3. Create a simple Product with attribute value 0 selected

### Steps to reproduce

1. Create configurable containing the simple product via REST Api

### Expected result

1. Configurable Product is created with the linked simple product

### Actual result

1. Exception thrown: Product with id \"%1\" does not contain required attribute \"%2\"."

The validation of the product links attributes from the configurable-product module is not type-safe which leads to a validation error when the value is (int)0
See: https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/ConfigurableProduct/Model/Plugin/ProductRepositorySave.php#L103

**Comment by davidverholen** [comment]
when it's confirmed I can provide a small pr just making the validation type safe or using the more suited hasData() here

**Comment by davidverholen** [comment]
update: it does not happen when saving the product in the backend because the problem is in a Plugin on the ProductRepository which is not used for saving the Product in the Adminhtml Save Controller

**Comment by davidverholen** [comment]
update: also some interesting behavior / bug: if you update the product via PUT request and do not specify type_id: configurable in the payload, the validation is not triggered since the product is not considered a configurable product here:  https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/ConfigurableProduct/Model/Plugin/ProductRepositorySave.php#L60

**Comment by magento-engcom-team** [comment]
Hi @davidverholen
We can not reproduce the issue as you describe.
Please share an install script, how do you add the attribute and share you api request.
Thanks

**Comment by davidverholen** [comment]
Ok, I've created a small module [config-bug.tar.gz](https://github.com/magento/magento2/files/1655766/config-bug.tar.gz)

- Installed it in a fresh Installation of the OS Edition 2.2.2
- Added an Integration with Admin permission
- Created a simple Product with Value Zero for "New Attribute"
- try to create a new configurable product for the simple via Rest API:
POST /V1/products
Payload: 
```json
{
  "product": {
    "sku": "config",
    "name": "config",
    "status": 1,
    "type_id": "configurable",
    "attribute_set_id":4,
    "extension_attributes": {
      "configurable_product_options": [
        {
          "attribute_id": 133,
          "label": "New Attribute",
          "values": [
            {
              "value_index": "0"
            }
          ]
        }
      ],
      "configurable_product_links":[
        1
      ]
    }
  }
}
```

Result:
```json
{"message":"Product with id \"%1\" does not contain required attribute \"%2\".","parameters":["1","new_attribute"]}
```

while the catalog_product_entity_int table also contains the correct value 0 for the attribute

**Comment by magento-engcom-team** [comment]
@davidverholen, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by davidverholen** [comment]
found this behavior in 2 more places: (since for some reason empty("0") evaluates to true in php):

When adding a product to cart with a "0" configuration chosen:
https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/ConfigurableProduct/Model/Product/Type/Configurable.php#L857

When having a product in cart with a "0" configuration chosen
https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/ConfigurableProduct/Model/Product/Type/Configurable.php#L955

as a first step, I think, it would be useful to extract the validation logic for the configurable options

**Comment by sdzhepa** [comment]
Hello @davidverholen 

Could you please create PR with fix for this issue to `2.4-develop`?

Thank you in advance!

**Comment by davidverholen** [comment]
at the time I created this issue the validation was implemented in several places  (instead of one central class for that responsibility).

Tbh I'm not sure if this is still an Issue, but if so, it should be fixed.
Unfortunately, currently I do not have the time to investigate further.

So it would be nice if someone could fix that

**Comment by engcom-Alfa** [comment]
Issue reproduced on the fresh 2.4-develop instance according to the steps, described in [comment ](https://github.com/magento/magento2/issues/13210#issuecomment-359773155)

**Actual Result:**
![screen](https://user-images.githubusercontent.com/51679138/71726000-6a084b80-2e3e-11ea-8899-4516ef9d2f1e.png)

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Alfa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-30199` were created

**Issue Available**: @engcom-Alfa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @davidverholen. Thank you for your report. 
The issue has been fixed in magento/magento2#29001 by @engcom-Charlie in 2.4-develop branch
Related commit(s):
 - [2e4b71cdcec16709a93fcfb22be7e482b3f66969](https://github.com/magento/magento2/commit/2e4b71cdcec16709a93fcfb22be7e482b3f66969)
 - [acd5177e9f442a3bed8f87e035d6d59bfc7a0f42](https://github.com/magento/magento2/commit/acd5177e9f442a3bed8f87e035d6d59bfc7a0f42)

The fix will be available with the upcoming 2.4.1 release.
